File: 1248.cpp
struct Solution {
    int atMostKOdd(const vector<int> &nums, const int k){
        if (k < 0){
            return 0;
        }
        int res = 0, oddCount = 0, s = 0;
        for (int e = 0; e < nums.size(); ++e){
            oddCount += nums[e] & 1;
            if (oddCount == k + 1){
                while ((nums[s] & 1) == 0){
                    ++s;
                }
                ++s;
                oddCount = k;
            }
            res += e - s + 1;
        }
        return res;
    }

    inline int numberOfSubarrays(vector<int>& nums, int k) {
        return this->atMostKOdd(nums, k) - this->atMostKOdd(nums, k - 1);
    }
};

File: sol/1248.py
from collections import deque

class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        res = 0
        q = deque((-1,))
        for i, x in enumerate(nums):
            if x & 1 == 0:
                continue
            if len(q) == k + 1:
                l = q.popleft()
                res += (q[0] - l) * (i - q[-1])
            q.append(i)
        if len(q) == k + 1:
            res += (q[1] - q[0]) * (len(nums) - q[-1])
        return res
